File: book_kit/src/main/java/net/lzbook/kit/data/bean/NovelLineBean.java
package net.lzbook.kit.data.bean;

import android.view.ViewGroup;

import net.lzbook.kit.constants.ReadConstants;

import android.graphics.Rect;
import android.text.TextUtils;

import java.io.Serializable;
import java.util.ArrayList;

/**
 * Created by yuchao on 2017/9/15 0015.
 */

public class NovelLineBean implements Serializable {
    private static final long serialVersionUID = 9173402120406347893L;
    private String lineContent;
    private float lineLength;
    //type:0-不是完整行 1-完整行
    private int type = -1;
    private boolean lastIsPunct;
    private ArrayList<Float> mArrLenths = new ArrayList<>();

    private int sequence;

    private String chapterName;

    private boolean lastPage;

    private int position;

    private float indexY;

    private ViewGroup adView;

    public ArrayList<Float> getArrLenths() {
        return mArrLenths;
    }

    public void setArrLenths(ArrayList<Float> arrLenths) {
        this.mArrLenths = arrLenths;
    }

    public boolean isLastIsPunct() {
        return lastIsPunct;
    }

    public void setLastIsPunct(boolean lastIsPunct) {
        this.lastIsPunct = lastIsPunct;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    public NovelLineBean() {
    }

    public NovelLineBean(String lineContent, float lineLength, int completeLine, boolean lastIsPunct, ArrayList<Float> arrLenths) {
        this.lineContent = lineContent;
        this.lineLength = lineLength;
        this.type = completeLine;
        this.lastIsPunct = lastIsPunct;
        initDrawIndex(arrLenths);
    }

    private void initDrawIndex(ArrayList<Float> arrLenths) {
        if (arrLenths == null || arrLenths.isEmpty() || TextUtils.isEmpty(lineContent)){
            return;
        }
        int length = lineContent.length();
        int charNum;
        if (lastIsPunct) {
            charNum = length - 2;
        } else {
            charNum = length - 1;
        }
        float marg = (ReadConfig.INSTANCE.getMWidth() - lineLength) / charNum;
        float star;
        if (!mArrLenths.isEmpty()){
            mArrLenths.clear();
        }
        for (int i = 0; i < length; i++) {
            star = ReadConfig.INSTANCE.getMLineStart() + arrLenths.get(i) + marg * i;
            char c = lineContent.charAt(i);
            if (i == length - 1 && isEndPunct(c)) {
                Rect rect = new Rect();
                ReadConfig.INSTANCE.getMPaint().getTextBounds(String.valueOf(c), 0, 1, rect);
                star -= marg;
                star -= rect.left;
                star += (ReadConstants.chineseWth / 2 - rect.width()) / 2;
            }
            mArrLenths.add(star);
        }
    }

    private boolean isEndPunct(char ch) {
        boolean isInclude = false;
        for (char c : ReadConstants.endPuncts) {
            if (ch == c) {
                isInclude = true;
                break;
            }
        }
        return isInclude;
    }

    public String getLineContent() {
        return lineContent;
    }

    public void setLineContent(String lineContent) {
        this.lineContent = lineContent;
    }

    public float getLineLength() {
        return lineLength;
    }

    public void setLineLength(float lineLength) {
        this.lineLength = lineLength;
    }

    public int getSequence() {
        return sequence;
    }

    public void setSequence(int sequence) {
        this.sequence = sequence;
    }

    public String getChapterName() {
        return chapterName;
    }

    public void setChapterName(String chapterName) {
        this.chapterName = chapterName;
    }

    public boolean isLastPage() {
        return lastPage;
    }

    public void setLastPage(boolean lastPage) {
        this.lastPage = lastPage;
    }

    public int getPosition() {
        return position;
    }

    public void setPosition(int position) {
        this.position = position;
    }

    public ViewGroup getAdView() {
        return adView;
    }

    public void setAdView(ViewGroup adView) {
        this.adView = adView;
    }

    @Override
    public String toString() {
        return "NovelLineBean{" +
                "lineContent=" + getLineContent() +
                '}';
    }

    public float getIndexY() {
        return indexY;
    }

    public void setIndexY(float indexY) {
        this.indexY = indexY;
    }
}

File: dy-reader/src/main/java/com/dy/reader/mode/NovelLineBean.kt
package com.dy.reader.mode

import com.dy.reader.setting.ReaderSettings

import android.graphics.Rect
import android.text.TextUtils
import com.dy.reader.constants.ReadConstants
import net.lzbook.kit.data.bean.ReadConfig.mWidth

import java.io.Serializable
import java.util.ArrayList

/**
 * Created by yuchao on 2017/9/15 0015.
 */

class NovelLineBean : Serializable {
    var lineContent: String? = null
    var lineLength: Float = 0.toFloat()
    //type:0-不是完整行 1-完整行
    var type = -1
    var isLastIsPunct: Boolean = false
    var arrLenths = ArrayList<Float>()

    var sequence: Int = 0

    var sequenceType: Int = 0

    var chapterName: String? = null

    var isLastPage: Boolean = false

    var position: Int = 0

    var indexY: Float = 0.toFloat()

    constructor() {}

    constructor(lineContent: String, lineLength: Float, completeLine: Int, lastIsPunct: Boolean, arrLenths: ArrayList<Float>?) {
        this.lineContent = lineContent
        this.lineLength = lineLength
        this.type = completeLine
        this.isLastIsPunct = lastIsPunct
        initDrawIndex(arrLenths)
    }

    private fun initDrawIndex(arrLenths: ArrayList<Float>?) {
        if (arrLenths == null || arrLenths.isEmpty() || TextUtils.isEmpty(lineContent)) {
            return
        }
        val length = lineContent!!.length
        val charNum: Int
        if (isLastIsPunct) {
            charNum = length - 2
        } else {
            charNum = length - 1
        }
        val marg = (ReaderSettings.instance.mWidth - lineLength) / charNum
        var star: Float
        if (!this.arrLenths.isEmpty()) {
            this.arrLenths.clear()
        }
        for (i in 0 until length) {
            star = ReaderSettings.instance.mLineStart + arrLenths[i] + marg * i
            val c = lineContent!![i]
            if (i == length - 1 && isEndPunct(c)) {
                val rect = Rect()
                ReaderSettings.instance.mPaint?.getTextBounds(c.toString(), 0, 1, rect)
                star -= marg
                star -= rect.left.toFloat()
                star += (ReadConstants.chineseWth / 2 - rect.width()) / 2
            }
            this.arrLenths.add(star)
        }
    }

    private fun isEndPunct(ch: Char): Boolean {
        var isInclude = false
        for (c in ReadConstants.endPuncts) {
            if (ch == c) {
                isInclude = true
                break
            }
        }
        return isInclude
    }

    override fun toString(): String {
        return "NovelLineBean{" +
                "lineContent=" + lineContent +
                '}'
    }

    companion object {
        private const val serialVersionUID = 9173402120406347893L
    }
}
